velox/connectors/clp/search_lib/ClpVectorLoader.cpp

```diff
@@ -14,16 +14,107 @@
+  Seconds,
+  Milliseconds,
+  Microseconds,
+  Nanoseconds
+};
+
+/// Estimates the precision of an epoch timestamp as seconds, milliseconds,
+/// microseconds, or nanoseconds.
+///
+/// This heuristic relies on the fact that 1 year of epoch nanoseconds is
+/// approximately 1000 years of epoch microseconds and so on. This heuristic
+/// can be unreliable for timestamps sufficiently close to the epoch, but
+/// should otherwise be accurate for the next 1000 years.
+///
+/// Note: Future versions of the clp-s archive format will adopt a
+/// nanosecond-precision integer timestamp format (as opposed to the current
+/// format which allows other precisions), at which point we can remove this
+/// heuristic.
+///
+/// @param timestamp
+/// @return the estimated timestamp precision
+template <typename T>
+auto estimatePrecision(T timestamp) -> TimestampPrecision {
+  constexpr int64_t kEpochMilliseconds1971{31536000000};
+  constexpr int64_t kEpochMicroseconds1971{31536000000000};
+  constexpr int64_t kEpochNanoseconds1971{31536000000000000};
+  auto absTimestamp = timestamp >= 0 ? timestamp : -timestamp;
+
+  if (absTimestamp > kEpochNanoseconds1971) {
+    return TimestampPrecision::Nanoseconds;
+  } else if (absTimestamp > kEpochMicroseconds1971) {
+    return TimestampPrecision::Microseconds;
+  } else if (absTimestamp > kEpochMilliseconds1971) {
+    return TimestampPrecision::Milliseconds;
+  } else {
+    return TimestampPrecision::Seconds;
+  }
+}
+
+auto convertToVeloxTimestamp(double timestamp) -> Timestamp {
+  switch (estimatePrecision(timestamp)) {
+    case TimestampPrecision::Nanoseconds:
+      timestamp /= Timestamp::kNanosInSecond;
+      break;
+    case TimestampPrecision::Microseconds:
+      timestamp /= Timestamp::kMicrosecondsInSecond;
+      break;
+    case TimestampPrecision::Milliseconds:
+      timestamp /= Timestamp::kMillisecondsInSecond;
+      break;
+    case TimestampPrecision::Seconds:
+      break;
+  }
+  double seconds{std::floor(timestamp)};
+  double nanoseconds{(timestamp - seconds) * Timestamp::kNanosInSecond};
+  return Timestamp(
+      static_cast<int64_t>(seconds), static_cast<uint64_t>(nanoseconds));
+}
+
+auto convertToVeloxTimestamp(int64_t timestamp) -> Timestamp {
+  switch (estimatePrecision(timestamp)) {
+    case TimestampPrecision::Nanoseconds:
+      break;
+    case TimestampPrecision::Microseconds:
+      timestamp *= Timestamp::kNanosecondsInMicrosecond;
+      break;
+    case TimestampPrecision::Milliseconds:
+      timestamp *= Timestamp::kNanosecondsInMillisecond;
+      break;
+    case TimestampPrecision::Seconds:
+      timestamp *= Timestamp::kNanosInSecond;
+      break;
+  }
+  int64_t seconds{timestamp / Timestamp::kNanosInSecond};
+  int64_t nanoseconds{timestamp - seconds * Timestamp::kNanosInSecond};
+  if (nanoseconds < 0) {
+    seconds -= 1;
+    nanoseconds += Timestamp::kNanosInSecond;
+  }
+  return Timestamp(seconds, static_cast<uint64_t>(nanoseconds));
+}
```

Comment:
```suggestion
auto convertToVeloxTimestamp(int64_t timestamp) -> Timestamp {
  int64_t seconds{timestamp};
  int64_t precisionDifference{1};
  switch (estimatePrecision(timestamp)) {
    case TimestampPrecision::Nanoseconds:
      break;
    case TimestampPrecision::Microseconds:
      precisionDifference = Timestamp::kNanosInSecond / Timestamp::kNanosecondsInMicrosecond;
      break;
    case TimestampPrecision::Milliseconds:
      precisionDifference = Timestamp::kNanosInSecond / Timestamp::kNanosecondsInMillisecond;
      break;
    case TimestampPrecision::Seconds:
      precisionDifference = Timestamp::kNanosInSecond / Timestamp::kNanosInSecond;
      break;
  }
  int64_t seconds /= precisionDifference;
  int64_t nanoseconds = timestamp % Timestamp::kNanosInSecond;
  if (nanoseconds < 0) {
    seconds -= 1;
    nanoseconds += Timestamp::kNanosInSecond;
  }
  return Timestamp(seconds, static_cast<uint64_t>(nanoseconds));
}
```
A possible way to avoid overflow permenantly

Reply:
maybe consider this way

Reply:
I think we would need to do
```
int64_t nanoseconds{timestamp % (Timestamp::kNanosInSecond / precisionDifference) *precisionDifference};
```
but yeah, this should work.

Should we update the implementation then, or is it fine to leave it?

Reply:
it would be good if we can update it in this PR, this would be the final comment for this PR